### ISSUE TITLE
Feature Proposal: SuperNumber

### ISSUE THREAD

**Issue by degecko** [laravel/ideas#1065]
I'm working on a class which takes a number and objectifies it so it can be used properly in an OOP environment.

Just to illustrate the idea, here are some examples which would become useful in real world applications:

```
$number = new SuperNumber(12000);
$someValue = 1e5;
$number->percentOf($someValue); // Returns 12, because 12000 is 12% of 100,000.

$number = new SuperNumber(15);
$someValue = 120;
$number->percentFrom($someValue); // Return 18, because 15% of 120 is 18.
```

It supports all the basic PHP math functions:

```
['abs', 'acos', 'acosh', 'asin', 'asinh', 'atan', 'atanh', 'ceil', 'cos', 'cosh', 'deg2rad', 'exp',
'expm1', 'floor', 'fmod', 'log10', 'log1p', 'log', 'min', 'max', 'rad2deg', 'round', 'sin', 'sinh',
'sqrt', 'tan', 'tanh', 'base_convert', 'bindec', 'decbin', 'dechex', 'decoct', 'hexdec', 'octdec',]

// E.g.:
$number = new SuperNumber(64);
$number->sqrt(); // Returns 8.
```

It also supports some additional functions, like add(), subtract(), multiply(), divide(), mod(), format() which is an alias for number_format, as well as a custom mutator to do whatever you want to the current value with a callable function.

I plan on adding more useful stuff.

And they're also chainable, which will look excellent in views for quick calculations.

This is a real world example from one of the apps that I'm working on.

I turned this:

`number_format($nextBonus->points - user()->pointsEarned, 0)`

Into this:

`$nextBonus->points->sub(user()->pointsEarned)->format()`

And this:

`number_format(user()->pointsEarned / $nextBonus->points * 100, 0)`

Into this:

`user()->pointsEarned->percentOf($nextBonus->points)`

Simply by initializing the SuperNumber class via the models' attribute getters.

```
class User extends Model
{
    // ...

    public function getPointsEarnedAttribute ($value) {
        return new SuperNumber((int) $value);
    }
}
```

Conclusion
--

I think this will be a great addition to Laravel's `$casts` model functionality, so I was thinking to create a pull request with this new class to be applied on a new casting type. But first, I'd like to see what you guys think of the idea. Maybe it's simply too much for just some silly numbers? Maybe it'll add to the load time?

I'd love to hear your opinions on this.

**Comment by NoelDavies** [comment]
This should be an external repository like https://github.com/Litipk/php-bignumbers, https://pear.php.net/package/Math_BigInteger or [BC Math](http://www.php.net/manual/en/book.bc.php)

**Comment by degecko** [comment]
@NoelDavies It will definitely be a separate repo, but I think it will be a great addition for Laravel too. :)

**Comment by sisve** [comment]
No, this sounds like a bad thing.

If you want a proper review of the idea, publish the code. What you've told us is just a vague idea about an Number object with a hexdec number, which sounds weird since we're sending an integer into the constructor. And you've not shown how your code handles floating point numbers, rounding, decimal precision, etc. All these are extremely important to get right, and impossible to say from the code and even numbers shown.

If we should go with something like this, then we should choose something with published code, published documentation and well-known.

**Comment by NoelDavies** [comment]
I agree with @sisve , build a package that implements the aforementioned functionality, or use an existing one and show it has support etc first.

Then look at where/how it can be implemented. Documentation, tests and support are a must for integration into a framework like this.

**Comment by degecko** [comment]
My apologies. It seems like it's been a misunderstanding. This is really a suggestion to find out if I should start implementing it.

Evidently, I cannot submit something that doesn't exist into the framework, and I'd have to build, test and document it first, but my question was if this would be a valid need, because I don't really want to spend time on something that's not needed or is already available through other packages.

**Comment by Kyslik** [comment]
>...valid need, because I don't really want to spend time on something that's not needed or is already available through other packages.

I guess you will never know until you actually do it and see if people use it, magic of open-source.

packagist.org is your friend

**Comment by sisve** [comment]
Don't do it. Use an existing package. There's no need for yet another library that handles numbers. If you're having issues with an existing library, contribute patches to that library instead of writing your own.
